### PR TITLE
Add `hilbert_series`

### PR DESCRIPTION
**Pull request by hannes14**
hilbert_series(I, Qt::PolyRing) wih the new Hilbert function algorithm for sideal and smodule

### CODE REVIEW

File: deps/src/ideals.cpp [hunk 1]
```diff
@@ -677,6 +677,22 @@ void singular_define_ideals(jlcxx::Module & Singular)
...
     });
+    Singular.method("scHilbPoly", [](ideal I, ring r, ring Qt) {
+        const ring origin = currRing;
+        rChangeCurrRing(r);
+        poly h=hFirstSeries0p(I,r->qideal,NULL,r,Qt);
+        rChangeCurrRing(origin);
+	return h;
+    });
+    Singular.method("scHilbPolyWeighted", [](ideal I, ring r, jlcxx::ArrayRef<int> weights, ring Qt) {
+        intvec *w = to_intvec(weights);
```
Comment (JohnAAbbott) [inline deps/src/ideals.cpp:688]: Assumes that `weights` has the correct size, and that all its entries are positive (and maybe not too big?)

File: src/ideal/ideal.jl [hunk 2]
```diff
@@ -1278,6 +1278,36 @@ function hilbert_series(I::sideal{spoly{T}}, w::Vector{<:Integer}) where T <: Ne
...
+    hilbert_series(I::sideal{spoly{T}}, w::Vector{<:Integer}) where T <: Nemo.FieldElem
+
+Return the polynomial $Q(t)$ of Qt where $\frac{Q(t)}{\prod_i (1-t^{w_i})}$
+is the Hilbert-Poincare series of $I$ for weights $\{w_i\}$. Each weight must be
+positive $w_i > 0$.
+"""
+function hilbert_series(I::sideal{spoly{T}}, w::Vector{<:Integer}, Qt::PolyRing) where T <: Nemo.FieldElem
+   I.isGB || error("Not a Groebner basis")
+   R = base_ring(I)
+   length(w) == nvars(R) || error("wrong number of weights")
```
Comment (JohnAAbbott) [inline src/ideal/ideal.jl:1304]: Is it useful in the error mesg to say how many weights were received, and how many were expected?

File: test/ideal/sideal-test.jl [hunk 3]
```diff
@@ -686,6 +686,11 @@ end
 
    @test ngens(std_hilbert(j, h, w, complete_reduction = true)) == 
          ngens(std(j, complete_reduction = true))
+    Qt,(t)= polynomial_ring(QQ, ["t"])
+    t=gens(Qt)[1]
```
Comment (JohnAAbbott) [inline test/ideal/sideal-test.jl:690]: I believe that TPTB prefer `gen(Qt,1)` over `gens(Qt)[1]`

File: test/ideal/sideal-test.jl [hunk 4]
```diff
@@ -686,6 +686,11 @@ end
...
+    I=Ideal(R,x,y,z)
```
Comment (JohnAAbbott) [inline test/ideal/sideal-test.jl:691]: This should probably be `I = ideal(R,[x,y,z])`

File: src/ideal/ideal.jl [hunk 5]
```diff
@@ -1278,6 +1278,36 @@ function hilbert_series(I::sideal{spoly{T}}, w::Vector{<:Integer}) where T <: Ne
...
 end
 
+@doc raw"""
+    hilbert_series(I::sideal{spoly{T}}, Qt::PolyRing) where T <: Nemo.FieldElem
+
+Return the polynomial $Q(t)$ as element of Qt where `Q(t)/(1-t)^nvars(base_ring(I))`
+is the Hilbert-Poincare series of $I$ for weights $(1, \dots, 1)$.
+"""
+function hilbert_series(I::sideal{spoly{T}}, Qt::PolyRing) where T <: Nemo.FieldElem
+   I.isGB || error("Not a Groebner basis")
```
Comment (JohnAAbbott) [inline src/ideal/ideal.jl:1288]: If it is the caller's responsibility to ensure that GB for `I` is available then this should be in the doc!
Maybe consider an optional/kw arg which the user can use to permit `hilbert_series` to compute a GB if necessary?

**Review by JohnAAbbott — COMMENTED** [review]
I have put some comments/suggestions in the code deltas

File: test/ideal/sideal-test.jl [hunk 6]
```diff
@@ -686,6 +686,11 @@ end
 
    @test ngens(std_hilbert(j, h, w, complete_reduction = true)) == 
          ngens(std(j, complete_reduction = true))
+   Qt,(t)= polynomial_ring(QQ, ["t"])
+   t=gen(Qt,1)
```
Comment (JohnAAbbott) [inline test/ideal/sideal-test.jl:690]: Is the assignment `t = gen(Qt,1)` superfluous?  I think line 698 already assigns the correct value to `t`, right?

**Review by JohnAAbbott — APPROVED** [review]
Not sure how I can actually try running this code; anyway, reading the code makes me believe that it is correct (under wholly reasonably assumptions about how Oscar-Singular interface works).  I believe line 690 in `sideal-test.jl` is superfluous (but harmless), and for me in an interactive Oscar session the syntax in line 691 produces an error (but works when `Ideal` is replaced by `ideal` (all lowercase)).

**Comment by hannes14** [comment]
Inconsistencies between Singular.jl and Oscar: Singular.jl has 7 variants of Ideal(..), but no ideal(..).
See for example the constructions in test/ideal/sideal-test.jl

**Review by JohnAAbbott — APPROVED** [review]
Approved: Hans has explained the parts which were unclear to me

**Review by JohnAAbbott — APPROVED** [review]
Thanks: neat workaround for one of julia's "foibles"

**Review by JohnAAbbott — APPROVED** [review]
Approved

File: test/ideal/sideal-test.jl [hunk 7]
```diff
@@ -686,6 +686,11 @@ end
 
    @test ngens(std_hilbert(j, h, w, complete_reduction = true)) == 
          ngens(std(j, complete_reduction = true))
+   Qt,_= polynomial_ring(QQ, ["t"])
+   t=gen(Qt,1)
```
Comment (fingolfin) [inline test/ideal/sideal-test.jl:690]:
```suggestion
   Qt,t= polynomial_ring(QQ, ["t"])
```

**Review by fingolfin — APPROVED** [review]

**Pull request merged**